feat: add adventurer metadata and tests

starknetdev opened the pull request:
Add the Adventurers metadata files.

Very similar approach to what was done for Realms metadata.

Waiting to have a better idea of `Adventurer.cairo` before altering that `tokenURI`.

starknetdev commented:
Added updates to `Adventurer.cairo`:

- Set module controller to get realm data
- Remove `setTokenURI`
- Alter `tokenURI` to calculate metadata from `metadata.cairo`

Adventurer items will only be in the metadata if they are equipped (`id` is not 0). Otherwise all other metadata is displayed even if the value is 0.

ponderingdemocritus commented:
nice work - it all looks logically correct with our current metadata. 

One structural thought I have been thinking about is the 'fixed' metadata that never changes. 'name' etc. Since we are packing and unpacking everything all the time, maybe it makes sense that we split this into another storage slot which we just fetch when needed instead of unpacking updating then repacking. 

So perhaps we think of metadata as:

- Static
- Dynamic

All static can exist unpacked - since it's only one storage write one time - then the savings over time will outstrip the constant packing and unpacking.

All dynamic needs to be packed since those slots will continue to change so the compute is cheaper

starknetdev commented:
@ponderingdemocritus 
As per your comments I have split the `AdventurerState` into `AdventurerStatic` and `AdventurerDynamic`. I have also fixed the metadata for items to reflect the `tokenId` and not `ItemIds`. Some points to note:

- Moved Order outside the `evolving stats` labelling and into `AdventurerStatic`. I assumed Order should only be set once?
- Changed the bit shifting and packed struct to correctly reflect the new `AdventurerDynamic` struct packing.
- Changed some of the `Adventurer.cairo` and `Loot.cairo` logic for the tokenId item lookup to work.
- Fix adventurer and metadata tests for new logic.

I have only made implementations for metadata, other `TODOs` to be done separately and liased with @loothero.

starknetdev commented:
@ponderingdemocritus per our discussion loot item metadata is implemented.

Note:
- Temporary settling game module controller ids have been added, Advernturer controller structure to be decided.
- Other `TODO`'s for Adventurers and Items to be done.

starknetdev commented:
@loothero @ponderingdemocritus metadata for loot items and adventurers is now updated and ready for the new logic.

**This PR should be merged after the logic PR**.

Metadata coverage:

Items:
* Description (Loot) -- Not Stored
* Name (contains:)
  * Name Prefix (Agony)* -- Not Stored
  * Name Suffix (Bite)* -- Not Stored
  * Name -- Not Stored
  * Suffix (Order of ...)* -- Not Stored
  * Max Greatness Suffix (+1)** -- Not Stored
* Image (Determined by item id) -- Not Stored
* Slot -- Not Stored
* Type -- Not Stored
* Material -- Not Stored
* Rank -- Not Stored
* Greatness -- Dynamic
* Created (unix timestamp) -- Not Stored
* XP -- Dynamic
* Adventurer (Name) -- Dynamic
* Bag -- Dynamic

Adventurer:
* Description (Adventurer) -- Not Stored
* Name -- Static
* Image (https://ipfs.io/ipfs + 2 `image_hash` parts)*** -- Static
* Race (1-10, see `contracts/loot/constants/adventurer.cairo` for list) -- Static
* Home Realm -- Static
* Order -- Static
* Birthdate (unix timestamp) -- Static
* Health -- Dynamic
* Level -- Dynamic
* Strength -- Dynamic
* Dexterity -- Dynamic
* Vitality -- Dynamic
* Intelligence -- Dynamic
* Wisdom -- Dynamic
* Charisma -- Dynamic
* Luck -- Dynamic
* XP -- Dynamic
* Weapon -- Dynamic
* Chest -- Dynamic
* Head -- Dynamic
* Waist -- Dynamic
* Feet -- Dynamic
* Hands -- Dynamic
* Neck -- Dynamic
* Ring -- Dynamic
* Status -- Dynamic
* Beast -- Dynamic

**Note:**
*Prefixes and suffix are displayed if greatness is greater or equal to 15. Currently these are just set to ids 1. Will there be a random generation event at greatness 15?
**Greatness suffix is displayed if greatness of item is greater or equal to 20.
***Adventurer `mint` takes parameters `image_hash_1` and `image_hash_2`. This is because stored hash of image is 46 characters, and so needs to be split and inputted as 2 felts.

starknetdev commented:
Second PR for implementing Adventurers module.

Nile commands (in order or execution):

* `nile mint_adventurer` - **Params:** `adventurer_token_id`, `race`, `home_realm`, `name`, `order`. **Logic:** Mint adventurer
* `nile get_adventurer` - **Params:** `adventurer_token_id`. **Logic:** Returns  adventurer data
* `nile mint_loot` - **Params:** None. **Logic:** Mints a random loot item
* `nile get_loot` - **Params:** `item_token_id`. **Logic:** Returns loot item data
* `nile set_loot` - **Params:** `item_token_id`, `item_id`, `greatness`, `xp`, `adventurer_token_id`, `bag`. **Logic:** Sets data to minted loot item
* `nile equip` - **Params:** `adventurer_token_id`, `item_token_id`. **Logic:** Equips item to correct slot
* `nile unequip` - **Params:** `adventurer_token_id`, `item_token_id`. **Logic:** Unequips item from correct slot
* `nile explore` -  **Params:** `adventurer_token_id`. **Logic:** Explores with adventurer to discover beast, 25% chance to discover beast, otherwise nothing is discovered
* `nile get_beast` - **Params:** `beast_token_id`. **Logic:** Returns data of beast
* `nile attack_beast` - **Params:** `beast_token_id`. **Logic:** Attacks a beast, if not killed will counteratttack
* `nile flee_from_beast` - **Params:** `beast_token_id`. **Logic:** 50% chance in flee or ambush, ambush deals damage to adventure

**NILE UPDATE**

Quite a large update to get to 0.11. **Currently need to pull latest main as bug in 0.11 version**:

 `pip install git+https://github.com/OpenZeppelin/nile.git`
 
  All deploy and nile command files need updated. Changes to make are updating to async, as well as inputs on functions such as `wrapped_send` and `logged_deploy` in `caller_invoke.py` and `deployer.py`. Please see file changes for more details.
